### PR TITLE
Add mapping type to collect the result in an array

### LINKED ISSUE #1321: I can collect a msgpack mapping result in an array
blocks #1232 
blocked by #1320

### PR DESCRIPTION
**Pull request by ThorbenLindhauer**
closes #1321

### CODE REVIEW

File: json-path/src/main/java/io/zeebe/msgpack/mapping/MsgPackTree.java [hunk 1]
```diff
@@ -211,4 +171,93 @@ public void mergeInto(MsgPackTree other) {
...
+  /**
+   * Creates or converts the addressed node to an array and appends the value as a new element.
+   * Replaces a previously existing non-array node completely.
+   */
+  public String appendToArray(
+      String parentId, String arrayNodeName, int documentId, int elementOffset, int elementLength) {
+
+    final String arrayNodeId = construct(parentId, arrayNodeName);
+
+    if (!hasNode(arrayNodeId)) {
```
Comment (Zelldon) [inline json-path/src/main/java/io/zeebe/msgpack/mapping/MsgPackTree.java]: Question out of interest:
What do you think is more readable?

```java
    if (hasNode(arrayNodeId))
    {
      if (!isArrayNode(arrayNodeId)) {
        clearChildren(arrayNodeId);
        convertToArrayNode(arrayNodeId);
      }
    }
    else {
      addArrayNode(parentId, arrayNodeName);
    }
```

Or:

```java
    if (!hasNode(arrayNodeId)) {
      addArrayNode(parentId, arrayNodeName);
    } else if (!isArrayNode(arrayNodeId)) {
      clearChildren(arrayNodeId);
      convertToArrayNode(arrayNodeId);
    }
```

File: json-path/src/main/java/io/zeebe/msgpack/mapping/MsgPackDocumentIndexer.java [hunk 2]
```diff
@@ -126,201 +101,144 @@
...
-
-      if (lastType != MsgPackType.ARRAY) {
-        parentsStack.push(nodeId);
-        nodeId = construct(nodeId, nodeName);
-      }
-    } else {
-      nodeId = createNodeId(nodeName);
-      isArrayValue = false;
-    }
+    switch (tokenContext.parsingMode) {
```
Comment (Zelldon) [inline json-path/src/main/java/io/zeebe/msgpack/mapping/MsgPackDocumentIndexer.java]: Alternative:

```java
   if (tokenContext.isMap) {
      if (mapEntryContext.IsKeyEntry) {
        mapEntryContext.currentKey = parseMapKey(currentValue);
        mapEntryContext.parsingMode = MapEntryParsingMode.VALUE;
      } else {
        parseValue(tokenContext, mapEntryContext.currentKey, position, currentValue);
        mapEntryContext.currentKey = "";
        mapEntryContext.parsingMode = MapEntryParsingMode.KEY;
      }
    } else {
      final int index = tokenContext.repetitions - tokenContext.remainingRepetitions - 1;
      final String key = Integer.toString(index);
      parseValue(tokenContext, key, position, currentValue);
    }
```

**Review by ChrisKujawa — COMMENTED** [review]
Looks good lets discuss in detail f2f.

**Comment by ThorbenLindhauer** [comment]
@Zelldon please see https://github.com/zeebe-io/zeebe/pull/1362/commits/5d5b630753aa9ab5f61276318e2e82f089bc35d8

**Review by ChrisKujawa — APPROVED** [review]

**Pull request merged**